File: server/api/overview.go
package api

import (
	"net/http"

	mw "github.com/cloudkarafka/cloudkarafka-manager/server/middleware"

	"github.com/cloudkarafka/cloudkarafka-manager/config"
	"github.com/cloudkarafka/cloudkarafka-manager/store"
)

type overviewVM struct {
	Version      string `json:"version"`
	Uptime       string `json:"uptime"`
	Brokers      int    `json:"brokers"`
	Topics       int    `json:"topics"`
	Partitions   int    `json:"partitions"`
	TopicSize    string `json:"topic_size"`
	Messages     int    `json:"messages"`
	Consumers    int    `json:"consumers"`
	MessageRates []int  `json:"message_rates"`
	BytesOut     []int  `json:"bytes_out"`
	BytesIn      []int  `json:"bytes_in"`
	ISRExpand    []int  `json:"isr_expand"`
	ISRShrink    []int  `json:"isr_shrink"`
}

func Overview(w http.ResponseWriter, r *http.Request) {
	var (
		user      = r.Context().Value("user").(mw.SessionUser)
		brokers   = store.DB.Brokers()
		topics    = topics(user.Permissions.DescribeTopic)
		consumers = store.DB.Consumers()
	)
	writeAsJson(w, overviewVM{
		Version:    config.Version,
		Uptime:     store.Uptime(),
		Brokers:    len(brokers),
		Topics:     len(topics),
		Consumers:  len(consumers),
		Partitions: store.Partitions(),
		TopicSize:  store.TotalTopicSize(),
		Messages:   store.TotalMessageCount(),
		BytesOut:   store.DB.SumBrokerSeries("bytes_out").All(),
		BytesIn:    store.DB.SumBrokerSeries("bytes_in").All(),
		ISRShrink:  store.DB.SumBrokerSeries("isr_shrink").All(),
		ISRExpand:  store.DB.SumBrokerSeries("isr_expand").All(),
	})
}

File: static/js/overview.js
/* global CloudKarafka Manager */
(function () {
  window.ckm = window.ckm || {}

  const url = '/api/overview'
  const raw = window.sessionStorage.getItem(cacheKey())
  let data = null
  let updateTimer = null

  if (raw) {
    try {
      data = JSON.parse(raw)
      if (data) {
        render(data)
      }
    } catch (e) {
      window.sessionStorage.removeItem(cacheKey())
      console.log('Error parsing data from sessionStorage')
      console.error(e)
    }
  }

  if (data === null) {
    update(render)
  }

  function cacheKey () {
    const user = window.sessionStorage.getItem('username')
    return url + '#' + user
  }


  function update (cb) {
    const headers = new window.Headers()
    ckm.http.request('GET', url, { headers }).then(function (response) {
      data = response
      try {
        window.sessionStorage.setItem(cacheKey(), JSON.stringify(response))
      } catch (e) {
        console.error('Saving sessionStorage', e)
      }
      render(response)
      if (cb) {
        cb(response)
      }
    }).catch(ckm.http.standardErrorHandler).catch(stop)
  }

  function render (data) {
    document.querySelector('#version').innerText = data.version
    const table = document.querySelector('#overview')
    if (table) {
      table.querySelector('.brokers').innerText = data.brokers
      table.querySelector('.topics').innerText = data.topics
      table.querySelector('.partitions').innerText = data.partitions
      table.querySelector('.topic_size').innerText = ckm.helpers.formatNumber(data.topic_size)
      table.querySelector('.messages').innerText = ckm.helpers.formatNumber(data.messages)
      table.querySelector('.consumers').innerText = data.consumers
      table.querySelector('.uptime').innerText = data.uptime
    }
  }

  function start (cb) {
    update(cb)
    updateTimer = setInterval(() => update(cb), 5000)
  }

  // Show that we're offline in the UI
  function stop () {
    if (updateTimer) {
      clearInterval(updateTimer)
    }
  }

  function get (key) {
    return new Promise(function (resolve, reject) {
      try {
        if (data) {
          resolve(data[key])
        } else {
          update(data => {
            resolve(data[key])
          })
        }
      } catch (e) {
        reject(e.message)
      }
    })
  }

  Object.assign(window.ckm, {
    overview: {
      update, start, stop, render, get
    }
  })
})()
